src/llms/openai.ts

```diff
@@ -0,0 +1,236 @@
+   * So 0.1 means only the tokens comprising the top 10% probability mass are considered.
+   *
+   * We generally recommend altering this or temperature but not both.
+   */
+  top_p?: number;
+  /**
+   * How many chat completion choices to generate for each input message.
+   */
+  n?: number;
+  /**
+   * Up to 4 sequences where the API will stop generating further tokens.
+   */
+  stop?: string | string[];
+  /**
+   * The maximum number of tokens to generate in the chat completion.
+   *
+   * The total length of input tokens and generated tokens is limited by the model's context length. Example Python code for counting tokens.
+   */
+  max_tokens?: number;
+  /**
+   * Number between -2.0 and 2.0.
+   *
+   * Positive values penalize new tokens based on whether they appear in the text so far, increasing the model's likelihood to talk about new topics.
+   */
+  presence_penalty?: number;
+  /**
+   * Number between -2.0 and 2.0.
+   *
+   * Positive values penalize new tokens based on their existing frequency in the text so far, decreasing the model's likelihood to repeat the same line verbatim.
+   */
+  frequency_penalty?: number;
+  /**
+   * Modify the likelihood of specified tokens appearing in the completion.
+   *
+   * Accepts a json object that maps tokens (specified by their token ID in the tokenizer) to an associated bias value from -100 to 100.
+   * Mathematically, the bias is added to the logits generated by the model prior to sampling. The exact effect will vary per model,
+   * but values between -1 and 1 should decrease or increase likelihood of selection; values like -100 or 100 should result in a ban
+   * or exclusive selection of the relevant token.
+   */
+  logit_bias?: { [key: string]: number };
+  /**
+   * A unique identifier representing your end-user, which can help OpenAI to monitor and detect abuse.
+   */
+  user?: string;
+}
+
+interface Choice {
+  message: Message;
+}
+
+interface ChatCompletionsResponse<T = Choice> {
+  choices: T[];
+}
+
+interface ContentMessage {
+  content: string;
+}
+
+interface RoleMessage {
+  role: string;
+}
+
+interface DoneMessage {
+  done: boolean;
+}
+
+type ChatCompletionsDelta = ContentMessage | RoleMessage | DoneMessage;
+
+interface ChatCompletionsChunk {
+  delta: ChatCompletionsDelta;
+}
+
+const isContentMessage = (message: any): message is ContentMessage => {
+  return message.content !== undefined;
+}
+
+const isDoneMessage = (message: any): message is DoneMessage => {
+  return message.done !== undefined;
+}
+
+/**
+  * Make a request to OpenAI's chat-completions API via the Grafana LLM plugin proxy.
+  */
+export async function chatCompletions(request: ChatCompletionsRequest): Promise<string> {
+  const response = await getBackendSrv().post<ChatCompletionsResponse>('/api/plugins/grafana-llm-app/resources/openai/v1/chat/completions', request, {
+    headers: { 'Content-Type': 'application/json' }
+  });
+  return response.choices[0].message.content;
```

Comment:
Should we support [function calling](https://platform.openai.com/docs/guides/gpt/function-calling) responses as well?) (e.g. check if message has a `function_call`, return otherwise). Happy to do this in a separate PR, need to make sure it works well with streaming response

Reply:
Good call, done in https://github.com/grafana/grafana-experimental/pull/72/commits/cef40b449c9ae1e1319f8b6ab0b02180afcf011b